firedrake/slate/slate.py

```diff
@@ -620,3 +607,21 @@ def __repr__(self):
     def _key(self):
         """Returns a key for hash and equality."""
         return (type(self), self.operands, self.actee)
+
+
+# Establishes levels of precedence for Slate tensors
+precedences = [
+    [Tensor],
+    [UnaryOp],
+    [Add, Sub],
+    [Mul, Action]
+]
+
+# Here we establish the precedence class attribute for a given
+# Slate tensor. Unary and binary operations will have higher
+# precedence over terminal Tensor objects. Similarly, tensor
+# action and multiplication will have higher precedence over
+# addition/subtraction
```

Comment:
Note that the compiler never asks for the precedence of a terminal (as expected, since there is no need).  Hence better to not even put a precedence property on terminals?